### ISSUE TITLE
DEFAULT 'now()' + column’s_TTL sometimes cause error “Code 40:..Checksums of parts don't match” during merges in ReplicatedMergeTree for version 20.10.7.4.

### ISSUE THREAD

**Issue by krebrikov** [ClickHouse/ClickHouse#21827]
1. Concise description of problem:

Sometimes merges in ReplicatedMergeTree table fail due to error: 
“_Code 40:… Checksums of parts don't match: uncompressed hash of compressed files doesn't match (version 20.10.7.4 (official build)). Data after merge is not byte-identical to data on another replicas..._” (there will be detailed error text below).

Assumed problem’s root cause is somehow related with 1 day column TTL used in this table, since the same environment+cluster+scheme doesn’t encounter «Checksums» merge errors when columns TTL is not used. 

Amount of errors depends on merge intensity. Daily load of about 30-35 billions of events (table is partitioned by YYYYMMDD) causes scores/several hundreds (by order of magnitude) of such errors per shard per day/partition.

In general it looks like parts that fail to merge due to considered error, are merged successfully later. 
That is in most cases this error is NOT regularly repeated for the data it affected once, thus this error does NOT prevent parts to be eventually merged. (I add this to point out that in a whole this error does not yet look like something critical what prevents merges, but rather some unsettling anomaly. However I'm not 100% certain it never escalates to something else).

2. Background/environment:

3 nodes Clickhouse cluster (version 20.10.7.4). 
Main fact table consists of 3 shards with 2 replicas (circle replication) based on local ReplicatedMergeTree tables all_events_v02.
(I know there are some arguments against circle replication, but we are already using this approach more than 2 years in production, while "Checksums" issue/error occured barely two weeks ago. So whatever drawbacks "circular replication" has - it is very unlikely to be related to root cause).
Data is loaded via: --> Null Table --> MV --> Distributed table all_events_upload --> local RMT tables all_events_v02. There are plenty other tables and MVs, but only this main fact table (containing >90% of database’ size) is affected.
Table is daily partitioned: ‘PARTITION BY toYYYYMMDD(end_datetime)’, where field end_datetime is event’s DateTime.
Field end_datetime is part of primary key.
Several columns have specific CODEC settings. 

Several fields have column TTL set as «TTL toDate(end_datetime) + toIntervalDay(1)» like:
`wdt` DateTime DEFAULT now() CODEC(Delta(4), ZSTD(1)) TTL toDate(end_datetime) + toIntervalDay(1),
`edt_ms` UInt32 DEFAULT 0 TTL toDate(end_datetime) + toIntervalDay(1),
`ch_host` LowCardinality(String) DEFAULT toString('') TTL toDate(end_datetime) + toIntervalDay(1),

(So on the next day after `end_datetime` new merges are supposed to clear contents of these columns - and this is when sometimes error occurs).

3. Error text:

```
2021.03.16 18:20:11.406853 [ 29317 ] {} <Debug> {databasename}.all_events_v02: Undoing transaction. Removing parts: 20210314_0_621_6.
2021.03.16 18:20:11.406939 [ 29317 ] {} <Error> {databasename}.all_events_v02: Code: 40, e.displayText() = DB::Exception: Checksums of parts don't match: uncompressed hash of compressed files doesn't match (version 20.10.7.4 (official build)). Data after merge is not byte-identical to data on another replicas. There could be several reasons: 1. Using newer version of compression library after server update. 2. Using another compression method. 3. Non-deterministic compression algorithm (highly unlikely). 4. Non-deterministic merge algorithm due to logical error in code. 5. Data corruption in memory due to bug in code. 6. Data corruption in memory due to hardware issue. 7. Manual modification of source data after server startup. 8. Manual modification of checksums stored in ZooKeeper. 9. Part format related settings like 'enable_mixed_granularity_parts' are different on different replicas. We will download merged part from replica to force byte-identical result.
```
( {database} is sometimes database related to first replica, and sometimes database related to second replica - due to circular/ring replication ).

I verified all 9 suggested causes mentioned in error text, but neither of them seems to be the case.
Custom settings (both system.settings and system.merge_tree_settings and system.replicated_merge_tree_settings) are identical on all nodes. No hardware issues takes place, there was no library upgrades. Etc.

However when column TTL for ReplicateMergeTree table (local all_events_v02) is not applied, and everything else is the same, - than there are no ‘Checksums’ errors.

4. Custom settings:
system.settings
┌─name─────────────────────┬─value──────┬─changed─┐
│ connect_timeout_with_failover_ms        	│ 500         	│       1 	│
│ connect_timeout_with_failover_secure_ms 	│ 550         	│       1 │
│ use_uncompressed_cache                  		│ 0           	│       1 │
│ load_balancing                          			│ in_order    	│       1 │
│ log_queries                             			│ 1           	│       1 │
│ http_connection_timeout                 		│ 10          	│       1 │
│ max_memory_usage                        		│ 75000000000 │       1 │
│ parallel_view_processing                		│ 1           	│       1 │
│ max_partitions_per_insert_block         	        │ 10          	│       1 │
│ materialize_ttl_after_modify            		│ 0           	│       1 │
│ max_memory_usage_for_all_queries        	│ 75000000000 │       1 │

system.replicated_merge_tree_settings
┌─name──────────────────────┬─value─┬─changed─┐
│ parts_to_delay_insert            			│ 150   		│       1 │
│ parts_to_throw_insert            			│ 600   		│       1 │
│ max_delay_to_insert              			│ 5     		│       1 │
│ max_suspicious_broken_parts      		│ 0     		│       1 │
│ min_merge_bytes_to_use_direct_io 		│ 0     		│       1 │

**Comment by krebrikov** [comment]
Thanks to Denny Crane, who provided an insight to the root cause of problem.

There is a DateTime field **wdt** in ReplicatedMergeTree table of my scheme that was subjected to "Code 40: Checksums..." error
`Code: 40, ... Checksums of parts don't match: uncompressed hash of compressed files doesn't match (version 20.10.7.4 (official build))...`

This field was defined as:
`wdt DateTime DEFAULT now() CODEC(Delta(4), ZSTD(1)) TTL toDate(end_datetime) + toIntervalDay(1)`

After I changed its DEFAULT expression to constant (DEFAULT 0), error no longer takes place:
`wdt DateTime DEFAULT 0 CODEC(Delta(4), ZSTD(1)) TTL toDate(end_datetime) + toIntervalDay(1)`

(TTL part of definition seems to be significant pre-condition for error. 
As far as I remember I already tried my scheme without column TTL and no merges caused 'Checksum of parts')

+++++++++++++++++++++++++++++++++

Interpretation:
Error was due to the result of DEFAULT expression ("now()") to be locally different on different replicas. 
Time of ZK's task creation or something else is intended to be used instead.
There was(is?)  some Issue/bug about it already (I'm not sure which particular one).
